velox/exec/tests/RowContainerTest.cpp

```diff
@@ -277,121 +278,298 @@ class RowContainerTest : public exec::test::RowContainerTestBase {
-    auto max = std::numeric_limits<T>::max();
-    auto nan = std::numeric_limits<T>::quiet_NaN();
-    auto inf = std::numeric_limits<T>::infinity();
-
-    facebook::velox::test::VectorMaker vectorMaker{pool_.get()};
-    VectorPtr values = vectorMaker.flatVectorNullable<T>(
-        {std::nullopt, max, inf, 0.0, nan, lowest, min});
-    int numRows = values->size();
-
-    SelectivityVector allRows(numRows);
-    DecodedVector decoded(*values, allRows);
-    std::vector<char*> rows(numRows);
-    std::vector<std::pair<int, char*>> indexedRows(numRows);
-    for (size_t i = 0; i < numRows; ++i) {
-      rows[i] = rowContainer->newRow();
-      rowContainer->store(decoded, i, rows[i], 0);
-      indexedRows[i] = std::make_pair(i, rows[i]);
+  static size_t countLeadingNulls(std::vector<std::optional<T>>& values) {
+    int32_t numNulls = 0;
+    for (auto i = 0; i < values.size(); ++i) {
+      if (!values[i].has_value()) {
+        ++numNulls;
+      } else {
+        break;
+      }
     }
 
-    std::vector<std::optional<T>> expectedOrder = {
-        std::nullopt, lowest, 0.0, min, max, inf, nan};
-    if (!ascending) {
+    return numNulls;
+  }
+
+  // Given a list of values sorted in ascending order with nulls first re-orders
+  // the elements to reverse the order and / or pu nulls last.
+  template <typename T>
+  void changeSortingOrder(
+      std::vector<std::optional<T>>& values,
+      bool ascending,
+      bool nullsFirst) {
+    if (ascending) {
       if (nullsFirst) {
-        std::reverse(expectedOrder.begin() + 1, expectedOrder.end());
+        // Nothing to do. 'values' are expected to be already sorted in
+        // ascending order with nulls first.
       } else {
-        std::reverse(expectedOrder.begin(), expectedOrder.end());
+        // Move the nulls from the start of the list to the end of the list.
+        const auto numNulls = countLeadingNulls(values);
+        for (auto i = 0; i < values.size() - numNulls; ++i) {
+          values[i] = values[i + numNulls];
+        }
+
+        for (auto i = values.size() - numNulls; i < values.size(); ++i) {
+          values[i].reset();
+        }
       }
     } else {
-      if (!nullsFirst) {
-        for (int i = 0; i < expectedOrder.size() - 1; ++i) {
-          expectedOrder[i] = expectedOrder[i + 1];
-        }
-        expectedOrder[expectedOrder.size() - 1] = std::nullopt;
+      if (nullsFirst) {
+        // Keeps nulls at the start of the list. Reverse the order of remaining
+        // non-null elements.
+        const auto numNulls = countLeadingNulls(values);
+        std::reverse(values.begin() + numNulls, values.end());
+      } else {
+        // Reverse the order of all elements.
+        std::reverse(values.begin(), values.end());
       }
     }
-    VectorPtr expected = vectorMaker.flatVectorNullable<T>(expectedOrder);
+  }
+
+  // A list of rows with unique zero-based indices.
+  using IndexedRows = std::vector<std::pair<vector_size_t, char*>>;
```

Comment:
Nit : Rename std::pair<vector_size_t, char*>> to RowAndIndex and the vector of it to RowAndIndexVector ?

Reply:
I used `IndexAndRowVector` because of the way the pair is defined (first, second).